### PR TITLE
add dobby.

### PR DESCRIPTION
**Pull request by Redbeanw44602**

### CODE REVIEW

File: packages/d/dobby/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    add_versions("2023.4.14", "0932d69c320e786672361ab53825ba8f4245e9d3")
+    
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "add-link-of-pthread.patch"), "e65f9b428e75db9d1994abf5695102c69a8ae17de36b13ef3d4f33fd6b361fd0")
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "fix-compile-on-lower-version-of-gcc.patch"), "632aad7d79e2afd9587089a39c3eb2b64a3750ab3c8954f04672c13abcddbbae")
+
+    -- build
+    add_configs("debug",   {description = "Enable debug logging.", default = false, type = "boolean"})
+    add_configs("example", {description = "Enable example build.", default = false, type = "boolean"})
+    add_configs("test",    {description = "Enable test build.",    default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:15]: we need not test, you can always set false.

File: packages/d/dobby/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    add_urls("https://github.com/jmpews/Dobby.git")
+
+    add_versions("2023.4.14", "0932d69c320e786672361ab53825ba8f4245e9d3")
+    
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "add-link-of-pthread.patch"), "e65f9b428e75db9d1994abf5695102c69a8ae17de36b13ef3d4f33fd6b361fd0")
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "fix-compile-on-lower-version-of-gcc.patch"), "632aad7d79e2afd9587089a39c3eb2b64a3750ab3c8954f04672c13abcddbbae")
+
+    -- build
+    add_configs("debug",   {description = "Enable debug logging.", default = false, type = "boolean"})
+    add_configs("example", {description = "Enable example build.", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:14]: and here

File: packages/d/dobby/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    add_urls("https://github.com/jmpews/Dobby.git")
+
+    add_versions("2023.4.14", "0932d69c320e786672361ab53825ba8f4245e9d3")
+    
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "add-link-of-pthread.patch"), "e65f9b428e75db9d1994abf5695102c69a8ae17de36b13ef3d4f33fd6b361fd0")
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "fix-compile-on-lower-version-of-gcc.patch"), "632aad7d79e2afd9587089a39c3eb2b64a3750ab3c8954f04672c13abcddbbae")
+
+    -- build
+    add_configs("debug",   {description = "Enable debug logging.", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:13]: debug is builtin configuraion.

File: packages/d/dobby/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    -- features
+    add_configs("near_branch",                       {description = "Enable near branch trampoline.",                              default = true,  type = "boolean"})
+    add_configs("full_floating_point_register_pack", {description = "Enables saving and packing of all floating-point registers.", default = false, type = "boolean"})
+
+    add_deps("cmake")
+    on_install("linux", "macosx", function (package)
+        function xmake_option(option)
+            return package:config(option) and "ON" or "OFF"
+        end
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:30]: we need not this wrapper. Please keep the configuration style consistent with other packages 

https://github.com/xmake-io/xmake-repo/blob/a4a8d20e32e602f7142721cdc9e87506715112bb/packages/l/libogg/xmake.lua#L25-L29

File: packages/d/dobby/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    set_license("Apache-2.0")
+
+    add_urls("https://github.com/jmpews/Dobby.git")
+
+    add_versions("2023.4.14", "0932d69c320e786672361ab53825ba8f4245e9d3")
+    
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "add-link-of-pthread.patch"), "e65f9b428e75db9d1994abf5695102c69a8ae17de36b13ef3d4f33fd6b361fd0")
+    add_patches("2023.4.14", path.join(os.scriptdir(), "patches", "fix-compile-on-lower-version-of-gcc.patch"), "632aad7d79e2afd9587089a39c3eb2b64a3750ab3c8954f04672c13abcddbbae")
+
+    -- plugins
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:12]: remove comments

File: packages/d/dobby/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    add_configs("symbol_resolver",             {description = "Enable symbol resolver plugin.",       default = true,  type = "boolean"})
+    add_configs("import_table_replacer",       {description = "Enable import table replacer plugin.", default = false, type = "boolean"})
+    add_configs("android_bionic_linker_utils", {description = "Enable android bionic linker utils.",  default = false, type = "boolean"})
+
+    -- features
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:17]: and here

File: packages/d/dobby/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    add_configs("near_branch",                       {description = "Enable near branch trampoline.",                              default = true,  type = "boolean"})
+    add_configs("full_floating_point_register_pack", {description = "Enables saving and packing of all floating-point registers.", default = false, type = "boolean"})
+
+    add_deps("cmake")
+    on_install("linux", "macosx", function (package)
+        local configs = {
+            "-DDOBBY_BUILD_EXAMPLE=OFF",
+            "-DDOBBY_BUILD_TEST=OFF",
+            "-DDOBBY_DEBUG="                     .. package:debug()                                     and "ON" or "OFF",
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:26]: use table.insert, add `()`, and rename to is_debug

```lua
table.insert(configs, "-DDOBBY_DEBUG=" .. (package:is_debug() and "ON" or "OFF"))
```

File: packages/d/dobby/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            "-DPlugin.SymbolResolver="           .. package:config("symbol_resolver")                   and "ON" or "OFF",
+            "-DPlugin.ImportTableReplace="       .. package:config("import_table_replacer")             and "ON" or "OFF",
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:28]: and here...

**Comment by Redbeanw44602** [comment]
Are we missing a clear coding style document?

File: packages/d/dobby/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,44 @@
+package("dobby")
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:1]: add `set_homepage`

**Comment by Redbeanw44602** [comment]
According to [platform_builder.py](https://github.com/Redbeanw44602/Dobby/blob/29712d5474877c04df20b04c51404d9e7371595a/scripts/platform_builder.py#L66), there may be some problems with dobby's shared library build, is there a way to temporarily disable the shared build option?

**Comment by star-hengxing** [comment]
```lua
if is_plat("macosx") then
    add_configs("shared", {description = "Build shared library.", default = false, type = "boolean", readonly = true})
end
```

File: packages/d/dobby/xmake.lua [hunk 10]
```diff
@@ -17,19 +18,34 @@ package("dobby")
...
         table.insert(configs, "-DPlugin.ImportTableReplace=" .. (package:config("import_table_replacer") and "ON" or "OFF"))
         table.insert(configs, "-DPlugin.Android.BionicLinkerUtil=" .. (package:config("android_bionic_linker_utils") and "ON" or "OFF"))
         table.insert(configs, "-DNearBranch=" .. (package:config("near_branch") and "ON" or "OFF"))
         table.insert(configs, "-DFullFloatingPointRegisterPack=" .. (package:config("full_floating_point_register_pack") and "ON" or "OFF"))
         table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
-        table.insert(configs, "-DCMAKE_SYSTEM_PROCESSOR=" .. package:targetarch())
+
+        if package:is_plat("android") then
+            local arch = package:arch()
+            local ndk = toolchain.load("ndk", {plat = package:plat(), arch = arch})
```
Comment (waruqi) [inline packages/d/dobby/xmake.lua:46]: local ndk = package:toolchain("ndk")

**Pull request merged**